### ISSUE TITLE
Getting site name and application name from IIS

### ISSUE THREAD

**Issue by MihaMarkic** [dotnet/aspnetcore#17069]
I would like to get both Application Name and Site Name from IIS when running under it.

In my case, I have an application running within same application pool but as different application names.

Like:
WebSite
 |-App1
 |-App2

where both App1 and App2 point to the same directory.

Hence I would like to differentiate my application instances based on IIS application name. Right now the only way would be to wait for first request and extract data from it's url. However, it'd be more elegant and more reliable if I had this information when the application starts.
Like for example IIS could pass this data as: 'WebSite/App1' and 'WebSite/App2'.

**Comment by davidfowl** [comment]
If you're using the same application pool, I'm assuming you're using the out of process model?

**Comment by MihaMarkic** [comment]
Hm, I'm in a planning phase. Good question. I suppose using in-process would mean sharing static data, right?

**Comment by davidfowl** [comment]
> Hm, I'm in a planning phase. Good question. I suppose using in-process would mean sharing static data, right?

In process doesn't support multiple apps in the same app pool, it just fails. Out of proc spawns 2 processes so there's no benefit to using the same app pool.

**Comment by MihaMarkic** [comment]
Yup, of course you are correct. Then out of process it is.

**Comment by Tratcher** [comment]
For out of process this is even more difficult. The only mechanism to flow information like this would be environment variables.

You can work around this today by setting your own configuration flags, correct?

FYI there's a very high bar for adding anything to out-of-process since it's a global component with high compat risks.

**Comment by MihaMarkic** [comment]
> You can work around this today by setting your own configuration flags, correct?

Not sure how would I do that? What configuration flags are you referring to? Environment variables would do the job for out of process I think.

**Comment by Tratcher** [comment]
Right, environment variables are a good option for you.

**Comment by vasont** [comment]
Having the same issue, but indirectly. In an ASP.net Core app, I need to find all bindings for the application, and I don't know what host it's running in. Old framework code I had to do this would retrieve System.Web.Hosting.HostingEnvironment.SiteName and then search IIS using Microsoft.Web.Administration.WebConfigurationManager.GetSection(null, null, "system.applicationHost/sites");, find the site by name, then get it's binding information, finally convert that to a list of binding names and return.

I'm doing all of this just to get bindings to dynamically add to a CORS allowed origins. @davidfowl is there a better way!?

**Comment by vasont** [comment]
I actually JUST stumbled across [CorsPolicyBuilder.SetIsOriginAllowedToAllowWildcardSubdomains Method](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.cors.infrastructure.corspolicybuilder.setisoriginallowedtoallowwildcardsubdomains?view=aspnetcore-3.1) which will suite my needs, but dang, this needs more attention for **real-world** _multi-tenant_ microservice developers.



> Having the same issue, but indirectly. In an ASP.net Core app, I need to find all bindings for the application, and I don't know what host it's running in. Old framework code I had to do this would retrieve System.Web.Hosting.HostingEnvironment.SiteName and then search IIS using Microsoft.Web.Administration.WebConfigurationManager.GetSection(null, null, "system.applicationHost/sites");, find the site by name, then get it's binding information, finally convert that to a list of binding names and return.
> 
> I'm doing all of this just to get bindings to dynamically add to a CORS allowed origins. @davidfowl is there a better way!?

**Comment by Tratcher** [comment]
@vasont getting binding information is a separate issue. These are available for IIS InProc via IServerAddressesFeature, but unfortunately not until the first request. You can see how the HttpsRedirectionMiddleware gets these values during the first request.

https://github.com/dotnet/aspnetcore/blob/master/src/Middleware/HttpsPolicy/src/HttpsRedirectionBuilderExtensions.cs
https://github.com/dotnet/aspnetcore/blob/master/src/Middleware/HttpsPolicy/src/HttpsRedirectionMiddleware.cs

**Comment by petarpetrovt** [comment]
We have similar problem migrating to asp.net core 2.0 (targeting net48). We use `HostingEnvironment.SiteName` and `Microsoft.Web.Administration.ServerManager` to get the current (first) site binding and start a background job (Hangfire) that requires the binding information.

**Comment by Tratcher** [comment]
In IIS-in-proc you can access the server variables and binding information. Note in-proc is only supported on core, not net4x.
https://github.com/dotnet/aspnetcore/blob/d7fc06af5e04a11a5ba90e9039f018f12c9f4a9d/src/Servers/IIS/IIS/samples/NativeIISSample/Startup.cs#L81-L84
https://docs.microsoft.com/en-us/iis/web-dev-reference/server-variables
Is `INSTANCE_NAME` what you're looking for?

https://github.com/dotnet/aspnetcore/blob/d7fc06af5e04a11a5ba90e9039f018f12c9f4a9d/src/Servers/IIS/IIS/samples/NativeIISSample/Startup.cs#L97-L101

**Comment by petarpetrovt** [comment]
@Tratcher yes `INSTANCE_NAME` will be sufficient but unfortunately we can't upgrade to core yet (.NET Remoting)

**Comment by petarpetrovt** [comment]
As an alternative i set custom environment variable for each site. Then use it to get the site bindings from the `ServerManager`.

```
appcmd.exe set config "mySite" -section:system.webServer/aspNetCore /+"environmentVariables.[name='TEST_SITE_NAME',value='mySite']" /commit:apphost
```

**Comment by jkotalik** [comment]
I don't believe we are going to add this to out of proc, and it seems like inproc handles all of these scenarios via server variables and serveraddress feature.
